### ISSUE TITLE
Multiple declarations of ‘:$$:$###’

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/singletons#200]
Yet another bug discovered when working on #160. This is only reproducible on `singletons-2.3` or later, since it relies on the fix to #172. This program:

```haskell
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE TypeOperators #-}
module Bug where

import Data.Singletons.TH

$(singletons [d|
      data ErrorMessage = ErrorMessage :$$: ErrorMessage

      ($$:) :: ErrorMessage -> ErrorMessage -> ErrorMessage
      x $$: y = x :$$: y
    |])
```
```
GHCi, version 8.2.0.20170523: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/rgscott/.ghci
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )

Bug.hs:10:3: error:
    Multiple declarations of ‘:$$:$###’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3
   |
10 | $(singletons [d|
   |   ^^^^^^^^^^^^^^...

Bug.hs:10:3: error:
    Multiple declarations of ‘:$$:$$###’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3
   |
10 | $(singletons [d|
   |   ^^^^^^^^^^^^^^...
```

This can be seen by inspecting the `-ddump-splices` output:

```
-- For (:$$:)
data (:$$:$$) (l_a6nl :: ErrorMessage_a6ne) (l_a6nk :: TyFun ErrorMessage_a6ne ErrorMessage_a6ne)
  = forall arg_a6nm. SameKind (Apply ((:$$:$$) l_a6nl) arg_a6nm) ((:$$:$$$) l_a6nl arg_a6nm) =>
    (:$$:$$###)
data (:$$:$) (l_a6nn :: TyFun ErrorMessage_a6ne (TyFun ErrorMessage_a6ne ErrorMessage_a6ne
                                                 -> GHC.Types.Type))
  = forall arg_a6no. SameKind (Apply (:$$:$) arg_a6no) ((:$$:$$) arg_a6no) =>
    (:$$:$###)

-- For ($$:)
data ($$:$$) (l_a6ns :: ErrorMessage_a6ne) (l_a6nr :: TyFun ErrorMessage_a6ne ErrorMessage_a6ne)
  = forall arg_a6nt. SameKind (Apply (($$:$$) l_a6ns) arg_a6nt) (($$:$$$) l_a6ns arg_a6nt) =>
    (:$$:$$###)
data ($$:$) (l_a6nu :: TyFun ErrorMessage_a6ne (TyFun ErrorMessage_a6ne ErrorMessage_a6ne
                                                -> GHC.Types.Type))
  = forall arg_a6nv. SameKind (Apply ($$:$) arg_a6nv) (($$:$$) arg_a6nv) =>
    (:$$:$###)
```

**Comment by RyanGlScott** [comment]
I suppose the solution here is to establish different conventions for kind-inference constructors depending on whether it's for a symbolic _value_ (whose kind-inference constructors have `:` prepended to them) or a symbolic _constructor_ (whose kind-inference constructors need not have `:` prepended all). For the former, we could use the suffix `###`, and for the latter, we could use the suffix `####`. Does that sound agreeable?

**Comment by RyanGlScott** [comment]
Hm, would we also need a similar convention for defunctionalization symbols? If you tweak the program slightly:

```haskell
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE TypeOperators #-}
module Bug where

import Data.Singletons.TH

$(singletons [d|
      data ErrorMessage = ErrorMessage :<>: ErrorMessage

      (<>:) :: ErrorMessage -> ErrorMessage -> ErrorMessage
      x <>: y = x :<>: y
    |])
```

Then when it errors, in addition to the aformentioned kind-inference constructor name clashes:

```
Bug.hs:10:3: error:
    Multiple declarations of ‘:<>:$###’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3
   |
10 | $(singletons [d|
   |   ^^^^^^^^^^^^^^...

Bug.hs:10:3: error:
    Multiple declarations of ‘:<>:$$###’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3
   |
10 | $(singletons [d|
   |   ^^^^^^^^^^^^^^...
```

You'll also get errors for the defunctionalization symbols themselves!

```
Bug.hs:10:3: error:
    Multiple declarations of ‘:<>:$’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3
   |
10 | $(singletons [d|
   |   ^^^^^^^^^^^^^^...

Bug.hs:10:3: error:
    Multiple declarations of ‘:<>:$$’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3
   |
10 | $(singletons [d|
   |   ^^^^^^^^^^^^^^...

Bug.hs:10:3: error:
    Multiple declarations of ‘:<>:$$$’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3
   |
10 | $(singletons [d|
   |   ^^^^^^^^^^^^^^...
```

**Comment by RyanGlScott** [comment]
This also applies to alphanumeric names:

```haskell
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE TypeOperators #-}
module Bug where

import Data.Singletons.TH

$(singletons [d|
      data Foo = MkFoo Bool Bool

      mkFoo :: Bool -> Bool -> Foo
      mkFoo x y = MkFoo x y
    |])
```

This gives duplicate defunctionalization symbols:

```
Bug.hs:10:3: error:
    Multiple declarations of ‘MkFooSym0’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3

Bug.hs:10:3: error:
    Multiple declarations of ‘MkFooSym1’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3

Bug.hs:10:3: error:
    Multiple declarations of ‘MkFooSym2’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3
```

As well as duplicate kind inference data constructors:

```
Bug.hs:10:3: error:
    Multiple declarations of ‘MkFooSym0KindInference’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3

Bug.hs:10:3: error:
    Multiple declarations of ‘MkFooSym1KindInference’
    Declared at: Bug.hs:10:3
                 Bug.hs:10:3
```

**Comment by goldfirere** [comment]
Ugh. These errors are tiresome (but admittedly unavoidable, given the way `singletons` conducts business). Any solution you can come up with is fine with me. :)

**Comment by RyanGlScott** [comment]
Good news: #230 solves the problems in https://github.com/goldfirere/singletons/issues/200#issue-235994123 and https://github.com/goldfirere/singletons/issues/200#issuecomment-309235079. It does _not_ solve the problem in https://github.com/goldfirere/singletons/issues/200#issuecomment-310866842, however, for separate reasons. I've opened #233 for this issue.

**Comment by RyanGlScott** [comment]
Fixed in #230.
